### PR TITLE
Add libravatar support, superseding gravatar

### PR DESCRIPTION
**Pull request by phikal**
[Libravatar](https://www.libravatar.org/) is a Gravatar alternative, that supports a certain degree of federation, but at the same time has a Gravatar fallback, if no profile picture was found.

I have _tried_ to use a [ruby gem](https://rubygems.org/gems/libravatar) that implements everything that is needed, while at the same time changing as little as possible on the user-end. Since I have never really done any Ruby, it might be that everything is wrong, but if I am given pointers, I'll be glad to fix the issues.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -333,19 +333,20 @@ def comments_deleted_count
   end
 
   def fetched_avatar(size = 100)
-    gravatar_url = "https://www.gravatar.com/avatar/" <<
-                   Digest::MD5.hexdigest(self.email.strip.downcase) <<
-                   "?r=pg&d=identicon&s=#{size}"
+    libravatar_url = Libravatar.new(:email => self.email,
```
[the remark this answers is not in the document]
Comment (phikal) [reply app/models/user.rb]: Thanks for the tip! I have force pushed a new commit that does what you say.

**Comment by pushcx** [comment]
I understand why you'd want to reduce our dependence on a random free service from a business with an open alternative, but there are significant downsides to this change. Libravatar doesn't even have an about page and there's nothing to build confidence that they'll still be operating a month, a year, or ten years from now. Gravatar has been running long enough that it'd be very surprising if Wordpress decided to shut it down. On the code side, this replaces a one-line string interpolation with a new gem that's hundreds of lines long; that's a lot of complexity to pick up. Finally, it's going to confuse users: if you visit [gravatar.com](https://www.gravatar.com) it explains what the service is about and why you'd want to use it; Libravatar doesn't have such a clear explanation and the subtle 'read more' link goes to a how-to page rather than explain much more.

**Comment by phikal** [comment]
If you're not interested, there's little I can do. But I'd still like to raise a few questions/conterarguments to what you mention:

> Libravatar doesn't even have an about page and there's nothing to build confidence that they'll still be operating a month, a year, or ten years from now

It is a federated service, but even so, according to [this](https://blog.libravatar.org/posts/Brave_new_Libravatar.org/) blog post, the project claims: "[...] Libravatar.org has run on Fedora Infrastructure Cloud. They are providing this service for free to us without any expiration date and with as much power as we need". Sure, they're not Wordpress, but it's not expected that they'll go under any time soon.

> On the code side, this replaces a one-line string interpolation with a new gem that's hundreds of lines long; that's a lot of complexity to pick up.

Sure, but that's the idea. You can't really implement a federated system in one line.

> Finally, it's going to confuse users: if you visit gravatar.com it explains what the service is about and why you'd want to use it

That's a social/marketing issue. One could mention it to the project, and it could be clarified.

**Comment by toastal** [comment]
Since these options use similar architectures and API. I 100% agree with the maintainer that it's pointless to bring in another library when the effective API is

```purescript
-- pseudocode assumes nilable origin (I don't know Ruby)
avatarURI :: forall r. { avatarOrigin : Maybe String, email : String | r } -> String
avatarURI user = origin <> "/avatar/" <> emailHash
  where
  -- defaults to Gravatar
  origin = fromMaybe "https://www.gravatar.com" user.avatarOrigin
  emailHash = md5Digest (toLower user.email)
```

I could see a new URL field in the user profile a text label for Libravatar `https://seccdn.libravatar.org` hint. 

```html
<label for="avatar_origin">
  <span>Avatar service origin: (e. g. Libravatar "https://seccdn.libravatar.org", defaults to Gravatar)</span>
  <input 
    type="url" 
    name="avatar_origin" 
    value="<%= user.avatarOrigin %>"
    placeholder="https://gravatar.com">
</label>
```

The same cache and invalidation of avatars can take place as production.

It might be a stretch too far to completely supplant Gravatar, but it seems equally disappointing for users to feel locked into any service at all -- especially one with privacy concerns (which I believe was the main purpose of the merge request). If an open URL field is opens too many security issues, then `<select>` with just Libravatar and Gravatar options and an enum could suffice, but wouldn't support the federated part of Libravatar (of course any malicious request could just ban the user).

**Pull request closed without merging**